astropy/io/fits/hdu/compressed.py

```diff
@@ -791,15 +848,12 @@ def _update_header_data(self, image_header,
 
         # Update the extension name in the table header
         if not name and 'EXTNAME' not in self._header:
-            name = 'COMPRESSED_IMAGE'
+            name = self._default_name
 
         if name:
             self._header.set('EXTNAME', name,
                              'name of this binary table extension',
                              after='TFIELDS')
-            self.name = name
-        else:
-            self.name = self._header['EXTNAME']
```

Comment:
`self.name` now sets `.header` which sets `._image_header` and `._header`. The assignment here was always suspect because it copies a EXTNAME value from `._header` and then ended up setting it back in `._header`.  It seemed always wrong. I can leave it in but that would lead to the _image_header getting the EXTNAME and that changes the checksum calculation (tests fail). Maybe that's the right thing to do but we should not be doing it if we are deciding to write COMPRESSED_IMAGE EXTNAME in (the default determined a few lines earlier). So I'm fine with retaining `self.name` and changing checksums in image_header so long as we don't do it for COMPRESSED_IMAGE.

Reply:
But, `_image_header` should also get the EXTNAME, otherwise it would not appear with `.header` ?

Not sure if that's related with those lines but I'm seeing some weird behavior:
```
In [5]: hdu = fits.CompImageHDU(data=np.arange(10), name='FOO')

In [6]: hdu._header['EXTNAME']
Out[6]: 'FOO'

In [7]: hdu.header  ## EXTNAME doesn't appear
Out[7]: 
XTENSION= 'IMAGE   '           / Image extension                                
BITPIX  =                   64 / array data type                                
NAXIS   =                    1 / number of array dimensions                     
NAXIS1  =                   10                                                  
PCOUNT  =                    0 / number of parameters                           
GCOUNT  =                    1 / number of groups                               

In [8]: hdu.name = 'BAR'

In [9]: hdu.header  ## OK
Out[9]: 
XTENSION= 'IMAGE   '           / Image extension                                
BITPIX  =                   64 / array data type                                
NAXIS   =                    1 / number of array dimensions                     
NAXIS1  =                   10                                                  
PCOUNT  =                    0 / number of parameters                           
GCOUNT  =                    1 / number of groups                               
EXTNAME = 'BAR     '           / extension name                                 

In [10]: hdu._header  ## Duplicate EXTNAME
Out[10]: 
...
ZVAL2   =                    4 / bytes per pixel (1, 2, 4, or 8)                
EXTNAME = 'BAR     '           / extension name                                 
EXTNAME = 'BAR     '           / extension name                                 
```

Reply:
The same happen with the default EXTNAME (`fits.CompImageHDU(data=np.arange(10))`), so I guess something to do with `CompImageHeader`.

Reply:
As a further comment the code before was doing:

if name is defined:
    set EXTNAME in _header to "name"
    set EXTNAME in _header via .name to "name"
else:
    set EXTNAME in _header via .name to "name"

The call of .name was completely irrelevant and image_header was not involved. Currently that is also true but because the image_header is not having EXTNAME set the `hdu.name` call will not return anything and will return the default of COMPRESSED_IMAGE. Previously hdu.name only looked at _header but now it looks at `.header`.  Option 1 is to do what I say above and update the `.header` and update the checksums. The other option is to allow `hdu.name` to first look in `.header` or else fallback to `._header` if they are out of sync.

Reply:
Yeah OK, well this is a mess :grin:, sorry for the trap. 
I didn't follow about the checksum issue, but if I understand correctly the checksum should change if EXTNAME is changed so this is a good thing isn't it ? (fixing another bug)

Reply:
Yes, so previously only `._header` was changing and not `.header` so yes, I was wondering if it was an actual bug that I fixed causing the checksums to change. I'll restore setting `.header` and fix the checksums but I'm fairly certain I don't want the COMPRESSED_IMAGE value to leak into `.header` since we spend all our time trying to make it disappear.

Reply:
I agree about COMPRESSED_IMAGE, it should only be added when writing a file if EXTNAME is not set.

Reply:
I have now done this. The checksums don't have to be fixed because in this specific example it was using the COMPRESSED_IMAGE default and I do not update .header in that situation.